### PR TITLE
LayerComposition refactor to simplify rendering to texture

### PR DESCRIPTION
**Pull request by mvaligursky**
Refactoring of the way LayerComposition builds up a list of cameras / layers to be rendered:
- The overall rendering order is driven by cameras, and not layers. All cameras are sorted by priority, and each camera renders all layers assigned to it in the order specified by layer ordering. When rendering to textures, this makes the engine to switch to each render target exactly one time per frame, instead of switching between render targets when walking layers, which was costing us fill rate heavy restore and resolve operations each time.
- This simplified the way rendering to texture can be set up, as seen in PR with examples for this: https://github.com/playcanvas/engine/pull/2757

**API changes**
- Layer.renderTarget is deprecated (still supported)
- Camera.RenderTarget is made public as preferred way to set up rendering to texture
- LayerComposition constructor takes optional name parameter, useful for debugging of multiple compositions

### CODE REVIEW

File: src/framework/components/camera/component.js
Comment (slimbuck) [inline src/framework/components/camera/component.js:182]: tiny point, but no need for layerComp variable.

**Review by slimbuck — APPROVED** [review]
Love it! Approving, but please get more approvals before merge.

File: src/framework/components/camera/component.js
Comment (mvaligursky) [reply src/framework/components/camera/component.js:182]: I agree, but I added it to make the code more readable as it's not clear from naming that scene.layers is LayerComposition.

**Comment by Maksims** [comment]
Just thinking through API change, here is case scenario of how I've used previously Layers for doing some effects:

Refraction: Have Refraction Layer, to which some entities were added (to be refracted), then this layer renders to texture. Which is then available for all the remaining layers to be rendered. So it was fairly simple to setup and manage, and easy to do shader stuff with it.

How it will be now? Will it be required to create layer with Refraction as before, but then render it in separate camera, which is set to render to texture? Sounds like extra unnecessary step to make in this case.

**Comment by mvaligursky** [comment]
@Maksims You can still use old API, but if you render multiple layers into the refraction, you'd get better performance to use extra camera. So:
Create Refraction camera, assign render target to it, and set it to render World layer (no new layer needed). This is the same as you'd render reflection texture .. create a camera that is plane reflected, and use that to render to texture.

**Comment by Maksims** [comment]
But this will not benefit from depth test, as geometry that is rendered in Refraction layer, already rendered. In two cameras, they will have own depth buffer, right?
I've actually used same approach for reflection as well, where I've flipped camera using events of before and after layer rendering. Instead of separate camera. It was simple small piece of code, no new entities, nothing else.

**Comment by mvaligursky** [comment]
If you set up both cameras to use the same render target (or null for framebuffer), they'd render to the same target and will benefit from depth being already rendered by the first camera, so that should be no problem.

**Comment by Maksims** [comment]
> If you set up both cameras to use the same render target (or null for framebuffer), they'd render to the same target and will benefit from depth being already rendered by the first camera, so that should be no problem.

Only refraction/reflection render should be in texture, other renders into back buffer.

**Comment by mvaligursky** [comment]
I'm not entirely sure what scenario you are talking about. Are you thinking about using single depth buffer when rendering into framebuffer and then use it again when rendering to refraction texture? If so, I'm not sure that was supported before? I looked at After The Flood few days ago, and that seems to have separate depth passes for both reflection, refraction and main cameras. Could you please give me more detailed example of how the rendering of those reflection / refraction layers you were doing to see how this could map to this new recommended way to set up rendering to textures.

This is how I would expect general water rendering to be setup:

- we'd have 2 layers for geometry object, World and Refraction.
- each geometry object would be in exactly one of these layers
- there will be RefractionCamera, rendering to Refraction RT, and only objects on Refraction Layer
- similarly, ReflectionCamera will render to Reflection RT, objects in World Layer (and maybe Refraction layer as well if they're not clean cut by the water level)
- main Camera will then render World and again maybe Refraction layer, using both Reflection and Refraction textures as inputs.
- all these 3 RTs (Reflection, Refraction, framebuffer) would use separate RT (so separate color and depth buffers), and can do depth prepass into depth if needed

File: src/scene/layer-composition.js
Comment (willeastcott) [inline src/scene/layer-composition.js:28]: So this is the default name for `GraphNode`s and `Material`s too. But other classes like `Texture` default to `null`. I kind of which I'd chosen `null` or empty string for `GraphNode` and `Material` TBH. Or at least enforced a case - probably all lower case. Kinda too late to rename the default for `GraphNode` and `Material` now though.... 😢

**Review by willeastcott — APPROVED** [review]

**Pull request merged**
